Add runtime register reconfiguration support and its corresponding demo SYN_flooding_protection

dyc2021 opened the pull request:
This is Yicun Duan's update.

Commit:
Add runtime register reconfiguration support and its corresponding demo SYN_flooding_protection

jiarong0907 commented on src/bm_sim/P4Objects.cpp:
```diff
@@ -2650,6 +2660,12 @@ P4Objects::map_json_value(const std::string &type, const std::string &name,
     }
+  } else if (type == "register_array") {
+    if (cfg_register_arrays_map.find(name) != cfg_register_arrays_map.end()) {
+      std::cout << "map_json_value: duplicated register_array name: " << name << std::endl;
```
Consider throwing an error or other ways here if this will cause fatal errors in the following processing.

jiarong0907 commented on src/bm_sim/P4Objects.cpp:
```diff
@@ -2698,6 +2714,13 @@ P4Objects::add_json_value(const std::string &type, const std::string &name,
     }
+  } else if (type == "register_array") {
+    if (cfg_register_arrays_map.find(name) != cfg_register_arrays_map.end()) {
+      std::cout << "add_json_value: duplicated register_array name: " << name << std::endl;
```
Similarly, consider other ways to handle this exception.

jiarong0907 commented on src/bm_sim/P4Objects.cpp:
```diff
@@ -2732,6 +2755,23 @@ P4Objects::remove_json_value(const std::string &type, const std::string &name) {
     }
+  } else if (type == "register_array"){
+    if (cfg_register_arrays_map.find(name) == cfg_register_arrays_map.end()) {
+      std::cout << "remove_json_value: no register_array name: " << name << std::endl;
```
Same comment here

jiarong0907 commented on src/bm_sim/P4Objects.cpp:
```diff
@@ -2732,6 +2755,23 @@ P4Objects::remove_json_value(const std::string &type, const std::string &name) {
+          newArray.append(cfg_register_array);
+        }
+      }
+      cfg_root["register_arrays"] = newArray;
```
2763-2769, is there a better way to remove the item from cfg_register_array directly? What's the reason for using this way?

jiarong0907 commented on src/bm_sim/P4Objects.cpp:
```diff
@@ -2770,6 +2810,12 @@ P4Objects::modify_json_value(const std::string &type, const std::string &name,
     }
+  } else if (type == "register_array") {
+    if (cfg_register_arrays_map.find(name) == cfg_register_arrays_map.end()) {
+      std::cout << "modify_json_value: cannot find register_array name: " << name << std::endl;
```
same comment here

jiarong0907 commented on src/bm_sim/P4Objects.cpp:
```diff
@@ -3236,6 +3337,60 @@ P4Objects::change_conditional_next_node_rt(const std::string &pipeline_name,
+    registers_new.reserve(new_register_array_size);
+    auto lock = register_array->unique_lock();
+    int i = 0;
+    while (i < new_register_array_size) {
```
Using for loop when the number of steps is determined.

jiarong0907 commented on src/bm_sim/P4Objects.cpp:
```diff
@@ -3236,6 +3337,60 @@ P4Objects::change_conditional_next_node_rt(const std::string &pipeline_name,
+      if (i < ori_register_array_size) {
+        registers_new.push_back(register_array->at(i));
+      } else {
+        registers_new.emplace_back(register_array->get_bitwidth(), register_array);
```
What does this do? 

BTW, the default value for an uninitialized register is 0.

jiarong0907 commented on src/bm_sim/P4Objects.cpp:
```diff
@@ -3271,6 +3426,82 @@ P4Objects::delete_match_table_rt(const std::string &pipeline_name,
+}
+
+void 
+P4Objects::rehash_register_array(const std::string &target_register_array,
```
The indent of this func is not good. Some lines are very long.

jiarong0907 commented on src/bm_sim/context.cpp:
```diff
@@ -169,6 +169,15 @@ Context::mt_runtime_reconfig(const std::string &json_file,
+        const std::string prefix = items[0].substr(0, 3);
+        const std::string actual_name = items[0].substr(4);
+        if (prefix != "new") {
+          std::cout << "Error: inserted register_array should only have prefix 'new_'\n";
```
exception handle

jiarong0907 commented on include/bm/bm_sim/P4Objects.h:
```diff
@@ -323,13 +330,27 @@ class P4Objects {
   void delete_match_table_rt(const std::string &pipeline_name,
                              const std::string &name);
+  void delete_register_array_rt(const std::string& name);
+  void rehash_register_array(const std::string& target_register_array,
```
This is a function with many parameters. Need to add comments talking about what each parameter does.

jiarong0907 replied:
first_hash_function, second_hash_function, third_hash_function, is a bit wired. Does this mean this function require only 3 hash functions or it could need more?

jiarong0907 reviewed (changes requested):
What's the difference between SYN_flooding_protection_old_utils and SYN_floolding_protection? Similarly, what's the difference between code and marked_code under SYN_floolding_protection? Do we really need all of them?

Commit:
Complete most parts of the tests, except rehash

Commit:
Complete all the tests and clean up the repository

Commit:
Fix some compile errors

Commit:
Pass all the tests and update the runtime_register_reconfig_readme.md

jiarong0907 commented on .gitignore:
```diff
@@ -109,3 +109,15 @@ vgcore*
+# idea configuration
+.idea/
+
+# files generated by checkinstall
```
> doc-pak/
backup-*
description-pak
*.deb 

Are they all generated by check install?

dyc2021 replied:
I'll rerun the checkinstall to check this

dyc2021 replied:
After running the checkinstall, git told me that there are following newly created files:

- backup-072320221847-pre-runtime-programmable.tgz
- description-pak
- doc-pak/
- runtime-programmable_1.0.0-1_amd64.deb

Under doc-pak/ there are:
```
.
|-- docs
|   |-- JSON_format.md
|   |-- p4dbg_user_guide.md
|   |-- performance.md
|   |-- runtime_CLI.md
|   `-- simple_switch.md
|-- LICENSE
|-- README.md
`-- VERSION

```

jiarong0907 replied:
Got it, thanks!

jiarong0907 commented on include/bm/bm_sim/P4Objects.h:
```diff
@@ -323,13 +330,25 @@ class P4Objects {
                                        const std::string &edge_name,
                                        const std::string &dst_node_name);
+  void change_register_array_size_rt(const std::string& name,
+                                     const std::string& register_array_size);
```
```suggestion
                                     const std::string& new_register_array_size);
```

jiarong0907 commented on include/bm/bm_sim/P4Objects.h:
```diff
@@ -323,13 +330,25 @@ class P4Objects {
+  void change_register_array_size_rt(const std::string& name,
+                                     const std::string& register_array_size);
+  void change_register_array_bitwidth_rt(const std::string& name,
+                                         const std::string& register_array_bitwidth);
```
```suggestion
                                         const std::string& new_register_array_bitwidth);
```

jiarong0907 commented on include/bm/bm_sim/P4Objects.h:
```diff
@@ -301,6 +301,10 @@ class P4Objects {
                                     const std::string &new_next_name);
+  const Json::Value build_register_array_json_value(int id,
+                                          const std::string &name,
+                                          const std::string &register_array_id,
```
There are two ids: id and register_array_id. What is the first one?

dyc2021 replied:
Now, it changes to:
```
const Json::Value build_register_array_json_value(int id,
                                          const std::string &name,
                                          const std::string &register_array_size,
                                          const std::string &register_array_bitwidth);
```

jiarong0907 commented on include/bm/bm_sim/P4Objects.h:
```diff
@@ -323,13 +330,25 @@ class P4Objects {
   void delete_match_table_rt(const std::string &pipeline_name,
                              const std::string &name);
+  void delete_register_array_rt(const std::string& name);
+  void rehash_register_array_rt(const std::string &target_register_array,
```
This function must have comments to explain the meaning of each argument and how it would work to achieve the functionality.

dyc2021 replied:
Add comments:
```
//! Rehashes the target register array referring to the provided parameters.
  //! 
  //! It's designed for the convenience of SYN_flooding_protection's demonstration. 
  //!
  //! Now, this rehash function is very straightforward. We assume that all historical recordings are accessible,
  //! and the rehash process is just to apply the hash function on these recordings again.
  //! 
  //! Runtime command:
  //! @code
  //! rehash register_array <target_register_array> 
  //! --according-to <recording_register_array> <recording_last_pos_register_array> <recording_counting_register_array> 
  //! --hash-function-for-counting <hash_function> 
  //! --hash-function-for-target <first_pos_hash_function> <second_pos_hash_function> ... <nth_pos_hash_function> 
  //! --reset <time_stamp_register_array>
  //!
  //! @endcode
  //!
  //! @param target_register_array the register array which accepts the rehash results.
  //! @param recording_register_array  the register array holding the historical recordings.
  //! (As for the demo SYN_flooding_protection, it stores the IP addresses of packets.)
  //! @param recording_last_pos_register the register which points to the end of recording register array.
  //! @param recording_counting_register_array the register array counting the number of accesses.
  //! (As for the demo SYN_flooding_protection, it counts the number of SYN packets going through the switch for each IP address.)
  //! @param hash_function_for_counting the hash function used by recording_counting_register_array. 
  //! (As for the demo SYN_flooding_protection, hash(IPv4_address) => a pos in recording_counting_register_array.)
  //! @param pos_hash_functions the hash functions for rehashing.
  //! (As for the demo SYN_flooding_protection, hash(IPv4_address) => a pos in target_register_array.)
  //! @param register_array_to_be_reset the register array to be reset after rehashing.
  //! (As for the demo SYN_flooding_protection, we reset the time stamp register array after each rehashing.)
```

jiarong0907 commented on include/bm/bm_sim/context.h:
```diff
@@ -182,13 +183,20 @@ class Context final {
                       const P4Objects::ForceArith &arith_objects);
 
+  RuntimeReconfigErrorCode
+  mt_runtime_reconfig_with_stream(std::istream* json_file_stream,
```
Why do we need this? This seems to be duplicated with the previous one? If you believe we do need this one, you should add a comment to explain it.

dyc2021 replied:
This is for tests;
Now, a comment is added to explain it:
```
// This function should only be explicitly called by test codes.
```

jiarong0907 commented on include/bm/bm_sim/stateful.h:
```diff
@@ -147,6 +154,12 @@ class RegisterArray : public NamedP4Object {
+
+  void set_bitwidth(int bitwidth) {
+    this->bitwidth = bitwidth;
+  }
```
```suggestion
  void set_bitwidth(int bitwidth) { this->bitwidth = bitwidth;  }
```

jiarong0907 commented on include/bm/bm_sim/stateful.h:
```diff
@@ -164,6 +177,11 @@ class RegisterArray : public NamedP4Object {
   // NOLINTNEXTLINE(runtime/references)
   void unlock(UniqueLock &lock) const { lock.unlock(); }
 
+  void reset_registers_with(std::vector<Register>& registers_new) { 
```
I believe this function needs a comment to explain what it does. I didn't understand what it does by only looking at its signature and implementation.

dyc2021 replied:
Add comments:
```
//! Exchange the underlying memory of this register array with registers_new.
//! This function is used to quickly update the content of a register array.
//! @param registers_new a vector of new registers to be swapped with the registers in this register array
```

jiarong0907 commented on include/bm/bm_sim/switch.h:
```diff
@@ -830,10 +832,62 @@ class SwitchWContexts : public DevMgr, public RuntimeInterface {
+
+    if (reconfig_return_code != static_cast<int>(RuntimeReconfigErrorCode::SUCCESS)) {
+      return reconfig_return_code;
+    }
```
Why do you cast both of the returned values to int? Why not compare the enum value directly?

dyc2021 replied:
Now, this function call:
`mt_runtime_reconfig_with_stream(0, &json_file_stream, &plan_file_stream);`
And in `mt_runtime_reconfig_with_stream`:
```
 RuntimeReconfigErrorCode reconfig_return_code = contexts.at(cxt_id).mt_runtime_reconfig_with_stream(json_file_stream, 
                                                                                                        plan_file_stream,
                                                                                                        get_lookup_factory(),
                                                                                                        required_fields,
                                                                                                        arith_objects);

    if (reconfig_return_code != RuntimeReconfigErrorCode::SUCCESS) {
      return static_cast<int>(reconfig_return_code);
    }

    if (output_json_file.empty()) {
      std::cout << "table reconfig successfully" << std::endl;
      return static_cast<int>(RuntimeReconfigErrorCode::SUCCESS);
    }

    std::ofstream ofs(output_json_file+".new", std::ios::out);
    if (!ofs) {
      BMLOG_ERROR("Error: cannot open output file: {}", output_json_file + ".new");
      return static_cast<int>(RuntimeReconfigErrorCode::OPEN_OUTPUT_FILE_FAIL);
    }

    contexts.at(cxt_id).print_runtime_cfg(ofs);

    std::cout << "table reconfig successfully" << std::endl;
    return static_cast<int>(RuntimeReconfigErrorCode::SUCCESS);
```

jiarong0907 commented on include/bm/bm_sim/switch.h:
```diff
@@ -830,10 +832,62 @@ class SwitchWContexts : public DevMgr, public RuntimeInterface {
+      contexts.at(cxt_id).mt_runtime_reconfig(json_file, plan_file,
+                                                  get_lookup_factory(),
+                                                  required_fields,
+                                                  arith_objects)
```
The indent of ); is a bit wired

dyc2021 replied:
Change this to:
`mt_runtime_reconfig_with_stream(0, &json_file_stream, &plan_file_stream);`

jiarong0907 commented on runtime_register_reconfig_readme.md:
```diff
@@ -0,0 +1,111 @@
+    - `delete register_array <name>`
+    - `change register_array_size <size>`
+    - `change register_array_bitwidth <width>`
+    - `rehash register_array <target_register_array> --according-to <recording_register_array> <recording_last_pos_register_array> <recording_counting_register_array> --hash-function-for-counting <hash_function> --hash-function-for-target <first_pos_hash_function> <second_pos_hash_function> ... <nth_pos_hash_function> --reset <time_stamp_register_array>`
```
```suggestion
    - `rehash register_array <target_register_array> 
        --according-to <recording_register_array> <recording_last_pos_register_array> <recording_counting_register_array> 
        --hash-function-for-counting <hash_function> 
        --hash-function-for-target <first_pos_hash_function> <second_pos_hash_function> ... <nth_pos_hash_function> 
        --reset <time_stamp_register_array>`

```

jiarong0907 commented on runtime_register_reconfig_readme.md:
```diff
@@ -0,0 +1,111 @@
+    - `delete register_array <name>`
+    - `change register_array_size <size>`
+    - `change register_array_bitwidth <width>`
+    - `rehash register_array <target_register_array> --according-to <recording_register_array> <recording_last_pos_register_array> <recording_counting_register_array> --hash-function-for-counting <hash_function> --hash-function-for-target <first_pos_hash_function> <second_pos_hash_function> ... <nth_pos_hash_function> --reset <time_stamp_register_array>`
```
This deserves detailed explanations of what it does. It's a very important function.

dyc2021 replied:
Explanations are added:

  (Please note that, this rehash function is very straight forward, and should only be used with SYN_flooding_protection demo)

        `target_register_array`: the register array which accepts the rehash results
        
        `recording_register_array`: the register array holding the historical recordings

        `recording_last_pos_register`: the register which points to the end of recording register array

        `recording_counting_register_array`: the register array counting the number of accesses

        `hash_function_for_counting`: the hash function used by recording_counting_register_array

        `pos_hash_functions`: the hash functions for rehashing

        `register_array_to_be_reset`: the register array to be reset after rehashing

jiarong0907 commented on runtime_examples/SYN_flooding_protection/app/error_rate_checker.py:
```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/env python3
```
The code here is good. Just want to mention that it might be better to call it normal user send, because error_rate_checker might lead people to think about the error rate of the defense.

dyc2021 replied:
The name is changed to `normal_user_send`

Commit:
Complete most of the tests for remaining parts

Please note that some tests are not passed, and some comments need to be addressed

Commit:
Fix some tests' codes, let them run without error

dyc2021 committed:
Commit the suggestion for include/bm/bm_sim/P4Objects.h

Co-authored-by: Jiarong Xing <42587491+jiarong0907@users.noreply.github.com>

dyc2021 committed:
Commit the suggestion for include/bm/bm_sim/P4Objects.h

Co-authored-by: Jiarong Xing <42587491+jiarong0907@users.noreply.github.com>

dyc2021 committed:
Commit the suggestion for include/bm/bm_sim/stateful.h

Co-authored-by: Jiarong Xing <42587491+jiarong0907@users.noreply.github.com>

dyc2021 committed:
Commit the suggestion for runtime_register_reconfig_readme.md

Co-authored-by: Jiarong Xing <42587491+jiarong0907@users.noreply.github.com>

Commit:
Address the comments, and update the readme

dyc2021 commented on runtime_register_reconfig_readme.md:
````diff
@@ -0,0 +1,111 @@
+    timeout
+    ...
+```
+This shows that, when the bloom filter's capacity is low (128 in this case), it might be possible for switch to prevent some legal users (such as `h2`) from getting in touch with the server (in this case `h3`, at `10.0.1.22`).
````
Change this line to:
```
This shows that, when the bloom filter's capacity is low (128 in this case), it suffers from high false positive rate and can intercept legitimate flows.
```

dyc2021 commented on include/bm/bm_sim/switch.h:
```diff
@@ -830,10 +832,62 @@ class SwitchWContexts : public DevMgr, public RuntimeInterface {
+                                                          get_lookup_factory(),
+                                                          required_fields,
+                                                          arith_objects)
+                                  );
```
Change this to:
```
 RuntimeReconfigErrorCode reconfig_return_code = contexts.at(cxt_id).mt_runtime_reconfig_with_stream(json_file_stream, 
                                                                                                        plan_file_stream,
                                                                                                        get_lookup_factory(),
                                                                                                        required_fields,
                                                                                                        arith_objects);
```

Commit:
Let mt_runtime_reconfig call mt_runtime_reconfig_with_stream

Commit:
Delete rehash related content

Commit:
Modify some comments

Commit:
Let helper functions return error code
